doc/sphinx/language/core/basic.rst

```diff
@@ -0,0 +1,428 @@
+  strings are identified with :production:`string`.
+
+Keywords
+  The following character sequences are reserved keywords that cannot
+  be used as identifiers::
+
+    _ Axiom CoFixpoint Definition Fixpoint Hypothesis IF Parameter Prop
+    SProp Set Theorem Type Variable as at by cofix discriminated else
+    end exists exists2 fix for forall fun if in lazymatch let match
+    multimatch return then using where with
+
+  Note that notations and plugins may define additional keywords.
+
+Other tokens
+  The set of tokens defined at any given time can vary because the
+  :cmd:`Notation` command can define new tokens.  A :cmd:`Require`
+  command may load more notation definitions, while the end of a
+  :cmd:`Section` may remove notations.  Some notations are defined in
+  the standard library (see :ref:`thecoqlibrary`) and are generally
+  loaded automatically at startup time.
+
+  Here are the character sequences that Coq directly defines as tokens
+  without using :cmd:`Notation`::
+
+    ! #[ % & ' ( () (bfs) (dfs) ) * ** + , - ->
+    . .( .. ... / : ::= := :> :>> ; < <+ <- <:
+    <<: <= = => > >-> >= ? @ @{ [ [= ] _
+    `( `{ { {| | |- || }
+
+  When multiple tokens match the beginning of a sequence of
+  characters, the longest matching token is used.  Occasionally you
+  may need to insert spaces to separate tokens.  For example, if ``~``
+  and ``~~`` are both defined as tokens, the inputs ``~ ~`` and ``~~``
+  generate different tokens, whereas if `~~` is not defined, then the
+  two inputs are equivalent.
+
+Basic notions
+-------------
+
+Glossary
+~~~~~~~~
+
+.. glossary::
+
+   term
+
+     Terms are syntactically valid objects constructed from the
+     following grammar.  When they are :term:`well-typed`, they
+     represent mathematical objects.
+
+     .. insertprodn term field_def
+
+     .. prodn::
+        term ::= @term_product_or_fun
+        | @term_let
+        | @term_if
+        | @term_fix
+        | @term_cofix
+        | @term100
+        term100 ::= @term_cast
+        | @term10
+        term10 ::= @term_application
+        | @one_term
+        one_term ::= @ @qualid_annotated
+        | @term1
+        term1 ::= @term_projection
+        | @term0 % @scope
+        | @term0
+        term0 ::= @qualid_annotated
+        | @sort
+        | @primitive_notations
+        | @term_evar
+        | @term_match
+        | @term_record
+        | @term_generalizing
+        | @term_ltac
+        | ( @term )
+        qualid_annotated ::= @qualid {? @univ_annot }
+
+     .. note::
+
+        Many commands and tactics use :n:`@one_term` rather than
+        :n:`@term`.  The former need to be enclosed in parentheses
+        unless they're very simple, such as a single identifier.  This
+        avoids confusing a space-separated list of terms or
+        identifiers with a :n:`@term_application`.
+
+   type
+
+     Types are a sub-class of the syntactical class of terms:
```

Comment:
> Think type and term should stay in Gallina spec language chapter.

Remember that the goal is to split this chapter apart and eventually remove it entirely. Furthermore, the new structure will alternate between description of term constructs and of commands. For instance, there is no point in describing the `match` constructs in term before having described the `Variant` command, nor the `fix` or `cofix` constructs before having described the `Inductive` and `CoInductive` commands.

> What does "syntactical class" add here? Is this used to include both well-typed and non-well-typed terms when "term" will generally be used to refer to well-type terms?
> 
> "class" itself may deserve a definition. Perhaps it's used in the literature rather than the common term "set"?

You are right that this is not very clear. Well-typed terms are also a subset of syntactically valid terms. The goal here was just to say that we cannot distinguish types purely on their syntax. I'll reformulate.

Reply:
Let me try again:
- the entries for "sentence" .. "attribute" are closely related and are presented in a logical reading order
- the entries for "term" and "type" are closely related and should appear together, but have little relationship to "sentence"...  It introduces many nonterminals that are not described at all here. 
 So it's a disconnect going from term/type to sentence...  I don't care where they end up, but _they don't belong here_.
- sentence... begins the top-down description of the grammar--an obvious entry point for the reader

Are you expecting to add a lot more glossary terms here?  I think that will not be very readable.  I expect you can have multiple `.. glossary::` directives in the document--no need to jam things together here.

Reply:
Terms, types, tactics and commands are the most essential Coq notions. Never mind that they are independent of one another, what is the problem of packing them in the same initial file? It has a logical meaning, which is that this is the base vocabulary you need to understand to go further.
Creating an arbitrary split between the notions of terms on the one hand, and commands on the other, was exactly what the Gallina specification language chapter did, and what I want to move away from because of what I said on `match` making no sense if you didn't describe `Variant` first, etc.

Reply:
If you think this is a bad idea, I'd like to know ASAP. In fact, I would have hoped to know from #12012 only, but I understand that this was a bit too abstract to comment on.

Reply:
> Terms, types, tactics and commands are the most essential Coq notions. Never mind that they are independent of one another, what is the problem of packing them in the same initial file?

Why not add something like the first sentence before the glossary?  As it is, the glossary seems a bit like a laundry list because the relationship between sentence and type is unclear/unstated.  We should be extra clear in this introductory material.

Also, there's no explanation of what terms and types are used for (unlike the others).

The long grammar for term is out of place; _none_ of it is explained here.  Command is fairly well described though it could be improved, e.g. doesn't cover selectors.

I think sentence... followed most logically after explaining lexical conventions.  At the least I'd put them at the top.

Reply:
I'll try to add much more explanations. I have to agree that the current state is not really readable.

Reply:
Check what I wrote.

Reply:
It's a little better.  I still think there's too much syntax for term here.  But this is not where and how we should introduce _term_ and especially _type_.  We don't yet have a proper high-level introduction to term and type and the relevant points from theory.  For example, that type checking is how you verify a proof and that type checking is just 2nd order unification.  The Curry-Howard correspondence. Also how CoC compares to other logical foundations.  Then this section can just focus on syntax, which seems more consistent with sentence...

This should go in the introduction chapter.  We should add this fairly soon.

We can organize it as it is for the time being and revisit this point when we have that introduction.